Implement PlayerViewDistanceChangeEvent

ShockedPlot7560 opened the pull request:
## Introduction
There is currently no simple way to recover a player's view distance change

### Relevant issues
* Fixes #4550 

## Changes
### API changes
* Add PlayerViewDistanceChangeEvent

### Behavioural changes
When a player requests a new distance, if it is different from the player's current distance, the event will be called.
The event will not be called when the player connecting because no view distance is established

## Tests
Make a simple plugin of this type: 
```php
class Main extends PluginBase implements Listener{

	public function onEnable() : void{
		$this->getServer()->getPluginManager()->registerEvents($this, $this);
	}

	public function onViewDistance(PlayerViewDistanceChangeEvent $event){
		$this->getServer()->broadcastMessage($event->getPlayer()->getName() . " request a different radius, " . $event->getNewDistance() . " instead of " . $event->getOldDistance());
	}
}
```
Go to your settings and change your rendering distance

ShockedPlot7560 committed:
implement PlayerViewDistanceChangeEvent

dktapps commented on src/network/mcpe/handler/InGamePacketHandler.php:
```diff
@@ -685,6 +686,13 @@ public function handleMapInfoRequest(MapInfoRequestPacket $packet) : bool{
 
 	public function handleRequestChunkRadius(RequestChunkRadiusPacket $packet) : bool{
+		if($this->player->getViewDistance() !== $packet->radius){
+			$ev = new PlayerViewDistanceChangeEvent($this->player, $packet->radius);
```
This is the wrong place. It won't be called if it's changed for reasons other than packet.

ShockedPlot7560 replied:
So in the Player class directly? This will group together all the way to change its view distance

dktapps commented on src/event/player/PlayerViewDistanceChangeEvent.php:
```diff
@@ -0,0 +1,50 @@
+	}
+
+	public function setRadius(int $radius) : void{
+		$this->radius = $radius;
```
This should have bounds checks.

dktapps replied:
I'm also doubtful this is a good idea.

dktapps commented on src/event/player/PlayerViewDistanceChangeEvent.php:
```diff
@@ -0,0 +1,50 @@
+ * Called when a player requests a different viewing distance than the current one.
+ */
+class PlayerViewDistanceChangeEvent extends PlayerEvent implements Cancellable{
+	use CancellableTrait;
```
I'm doubtful that this should be cancellable. It will cause bugs on the client unless synced back.

ShockedPlot7560 replied:
Indeed, when the player connects, if a plugin cancels the event, it will be blocked

JavierLeon9966 replied:
What if the player changes the view distance while in-game?

ShockedPlot7560 replied:
If the player is already logged in, this was not going to take their request. However, making this event cancellable is not the right solution because it can block some behavior.

ShockedPlot7560 committed:
apply suggestion
- remove cancellable event
- move behavior to Player class
- add old radius to the event

dktapps commented on src/event/player/PlayerViewDistanceChangeEvent.php:
```diff
@@ -0,0 +1,48 @@
+	}
+
+	public function getRadius() : int{
+		return $this->radius;
```
This should probably be renamed to `getNewRadius()` since we're having a `getOldRadius()` here.

ShockedPlot7560 replied:
For sure, it's better indeed

ShockedPlot7560 committed:
rename getRadius to getNewRadius

dktapps commented on src/player/Player.php:
```diff
@@ -485,6 +487,11 @@ public function setViewDistance(int $distance) : void{
+		if($this->viewDistance !== $oldViewDistance){
+			$ev = new PlayerViewDistanceChangeEvent($this, $this->viewDistance, $oldViewDistance);
+			$ev->call();
+		}
```
I think this should be called before the view distance change, rather than after. If we decide to make it cancellable later on, we don't want to subtly break plugin assumptions about `Player->getViewDistance()` during the event.

dktapps commented on src/event/player/PlayerViewDistanceChangeEvent.php:
```diff
@@ -0,0 +1,48 @@
+	protected int $newRadius;
+	protected int $oldRadius;
+
+	public function __construct(Player $player, int $newRadius, int $oldRadius){
```
Weird parameter order?

ShockedPlot7560 committed:
fix weird parameter order

ShockedPlot7560 committed:
change the event call order

ShockedPlot7560 committed:
fix CS

dktapps commented:
I think this should use the `distance` terminology consistently, rather than `radius`.

dktapps commented on src/event/player/PlayerViewDistanceChangeEvent.php:
```diff
@@ -0,0 +1,48 @@
+		$this->newRadius = $newRadius;
+	}
+
+	public function getNewRadius() : int{
```
Documentation should be added to indicate that this number is in chunks.

dktapps commented on src/event/player/PlayerViewDistanceChangeEvent.php:
```diff
@@ -0,0 +1,48 @@
+		return $this->newRadius;
+	}
+
+	public function getOldRadius() : int{
```
Same again here.

ShockedPlot7560 committed:
use of distance instead of radius terminology

ShockedPlot7560 committed:
add PHPdoc for event

dktapps commented on src/event/player/PlayerViewDistanceChangeEvent.php:
```diff
@@ -29,20 +29,27 @@
-	public function getNewRadius() : int{
-		return $this->newRadius;
+	/**
+	 * Returns an int corresponding to the number of chunks forming the new view distance
```
This is super misleading.

dktapps replied:
```suggestion
	 * Returns the new view radius, measured in chunks.
```

dktapps commented on src/event/player/PlayerViewDistanceChangeEvent.php:
```diff
@@ -29,20 +29,27 @@
-		return $this->oldRadius;
+	/**
+	 * Returns an int corresponding to the number of chunks forming the previous view distance, before change
+	 * If the player connects to the server, it will return -1
```
```suggestion
	 * A value of -1 means that the player has just connected and did not have a view distance before this event.
```

ShockedPlot7560 replied:
As this is very clear, I have not been able to express it more clearly.

dktapps replied:
"if the player connects" is unclear. Every player connects to the server ... does that mean it will always return -1? No.

dktapps commented on src/event/player/PlayerViewDistanceChangeEvent.php:
```diff
@@ -29,20 +29,27 @@
-	public function getOldRadius() : int{
-		return $this->oldRadius;
+	/**
+	 * Returns an int corresponding to the number of chunks forming the previous view distance, before change
```
```suggestion
	 * Returns the old view radius, measured in chunks.
```

ShockedPlot7560 committed:
Apply suggestions for PHPdoc

Co-authored-by: Dylan T. <odigiman@gmail.com>
